tests/initcoder_test.go

```diff
@@ -0,0 +1,170 @@
+	"github.com/stretchr/testify/require"
+)
+
+// this constant comes from  https://eips.ethereum.org/EIPS/eip-3860#parameters
+const MAX_INIT_CODE_SIZE uint64 = 49152
+const sufficientGas = uint64(100_000)
+
+func TestInitCodeSizeLimitAndMetered(t *testing.T) {
+	require := require.New(t)
+
+	net, err := StartIntegrationTestNet(t.TempDir())
+	require.NoError(err)
+	defer net.Stop()
+
+	// Deploy the invalid start contract.
+	contract, receipt, err := DeployContract(net, initcode.DeployInitcode)
+	require.NoError(err)
+	require.Equal(types.ReceiptStatusSuccessful, receipt.Status, "failed to deploy contract")
+
+	// deploy measureGasAndAssign to get cost of deploying a contract without create.
+	_, gasUsed := createAndCheckContractWithCodeLenAndGas(t, net, contract, contract.MeasureAssignGasCost, 0, 0)
+
+	// -- using CREATE instruction
+	const wordCostCreate uint64 = 2
+	var gasForCreate uint64 = 32000 + gasUsed
+	t.Run("create", func(t *testing.T) {
+		testForVariant(t, net, contract, contract.CreateContractWith, gasForCreate, wordCostCreate)
+	})
+
+	// -- using CREATE2 instruction
+	// create2 has extra costs for hashing, this is explained in
+	// https://eips.ethereum.org/EIPS/eip-3860 and reflected in evm.codes calculator.
+	const wordCostCreate2 uint64 = wordCostCreate + 6
+	var gasForCreate2 uint64 = gasForCreate + 44
+	t.Run("create2", func(t *testing.T) {
+		testForVariant(t, net, contract, contract.Create2ContractWith, gasForCreate2, wordCostCreate2)
+	})
+
+	t.Run("create transaction", func(t *testing.T) {
+		t.Run("charges depending on the init code size", func(t *testing.T) {
+			// transactions charge 4 gas for each zero byte in data.
+			const zeroByteCost uint64 = 4
+			// create a transaction with 1 word of code.
+			receipt1, err := runTransactionWithCodeSizeAndGas(t, net, 1, sufficientGas)
+			require.NoError(err)
+			require.Equal(types.ReceiptStatusSuccessful, receipt1.Status,
+				"failed on transfer to empty receiver with valid code")
+
+			// createa a transaction with 2 byte of code.
+			receipt2, err := runTransactionWithCodeSizeAndGas(t, net, 2, sufficientGas)
+			require.NoError(err)
+			require.Equal(types.ReceiptStatusSuccessful, receipt2.Status,
+				"failed on transfer to empty receiver with valid code")
+
+			difference := receipt2.GasUsed - receipt1.GasUsed
+			require.Equal(difference, zeroByteCost,
+				"gas difference between 1 and 2 words should be 4, instead got", difference)
+		})
+
+		t.Run("aborts with init code size larger than MAX_INITCODE_SIZE", func(t *testing.T) {
+			// as specified in https://eips.ethereum.org/EIPS/eip-3860#rules,
+			// this is similar to transactions considered invalid for not meeting the intrinsic gas cost requirement.
+			_, err := runTransactionWithCodeSizeAndGas(t, net, MAX_INIT_CODE_SIZE+1, sufficientGas)
+			require.ErrorContains(err,
+				"intrinsic gas too low", "unexpectedly succeeded to create contract with init code length greater than MAX_INITCODE_SIZE")
```

Comment:
If the error is "intrinsic gas too low" then the transaction was not stopped by the code-size limit but by insufficient gas. I do not think that both of these errors should be accepted here. Only the second.

Is `require.ErrorContains` a either or or a all-of-those condition?

Reply:
as the comment above says, the first 2 rules in the EIP say that intrinsic gas error is the expected one. Or at least that was my interpretation. 

` ErrorContains(theError error, contains string, msgAndArgs ...interface{})` is the signature, I think this answers your quesiton.